Wg/migrator dev docs (#135)
sourcegraph/docs: docs/dev/background-information/sql/upgrades_and_migrator_design.mdx

# Upgrades and Migrator Design

> *This doc is based on a preface for [RFC 850 WIP: Improving The Upgrade

Reviewer: ```suggestion
<p className="subtitle">This doc is based on a preface for [RFC 850 WIP: Improving The Upgrade
```

Experience](https://docs.google.com/document/d/1ne8ai60iQnZfaYuB7QLDMIWgU5188Vn4_HBeUQ3GASY/edit)
and acts as a summary of the current design of our upgrade and database schema
management design. Subsequently additions have been made. The docs initial aim was to provide general information about the
relationship of our migrator service to our deployment types and

> migrator's dependencies during the release process.*

Reviewer: ```suggestion
migrator's dependencies during the release process.</p>
```

## Migrator Overview

> The `migrator` service is a short lived container responsible for managing 

Reviewer: ```suggestion
The `migrator` service is a short-lived container responsible for managing 
```

> Sourcegraph's databases (`pgsql` (*also refered to as frontend*), `codeintel-db`, 

Reviewer: ```suggestion
Sourcegraph's databases (`pgsql` (*also referred to as frontend*), `codeintel-db`, 
```

and `codeinsights-db`), and running schema migrations during startup and upgrades.

> Its design accounts for a variety of unique characteristics of 

Reviewer: ```suggestion
Its design accounts for  various unique characteristics of 
```

versioning and database management at Sourcegraph. Specifically graphical 
schema migrations, out of band migrations, and periodic schema migration squashing.

Sourcegraph utilizes a [directed acyclic graph](https://github.com/sourcegraph/sourcegraph/pull/30664) 

> of migration definitions, rather than a linear chain. In Sourcegraphs early days when schema migrations 

Reviewer: ```suggestion
of migration definitions, rather than a linear chain. In Sourcegraph's early days when schema migrations 
```

were applied linearly, schema changes were frequent enough that schema changes generally conflicted with
the master branch by the time a PR passed CI. Moving to a graph of migrations means, devs won't need to 
worry about other teammates concurrent schema changes unless they are working on the same table.

Similarly [squashing](#squashing-migrations) of schema migrations into a root definition reduced the number of migrations run on startup,
alleviating a common issue in which frequent transaction locks caused failed migration on Sourcegraph startup. 

> You can learn more in our [migrations overview docs](./migrations_overview.mdx#in-band-migrations). 

Reviewer: ```suggestion
You can learn more in our [migrations overview docs](/migrations_overview#in-band-migrations). 
```

> Information on Out of bound migrations can also be found there.

Reviewer 2: ```suggestion
Information on out of band migrations can also be found there.
```

> Information on Out of bound migrations can also be found there.

Reviewer: ```suggestion
Information on out of bound migrations can also be found there.
```

### CLI design

Migrator is designed with a CLI tool interface -- taking various commands to alter
the state of the database and apply necessary schema changes. This
design was initially implemented as a tool for TS team members to assist
in multiversion upgrades and because it could easily be included over
multiple deployment methods as a containerized process run separately
from the frontend during startup ([which originally caused
issues](https://about.sourcegraph.com/blog/introducing-migrator-service)).
This replaced earlier go/migrate based strategies which ran in the
frontend on startup. While the migrator can operate as a CLI tool, it's
containerized as if it was another application which allows it to be run
as an initContainer (in Kubernetes deployments) or as a dependent
service (for Docker Compose deployments) to ensure that the necessary
migrations are applied before startup of the application proper.
Check out [RFC 469](https://docs.google.com/document/d/1_wqvNbEGhMzu45CHZNrLQYys7gl0AtVpLluGp4cbQKk/edit#heading=h.ddeuyk4t99yx) to learn more.

### Important Commands

The most important migrator commands are `up` and `upgrade`, with a notable mention to `drift`:

> -   **Up** : The default command of `migrator`, `up` ensures that

Reviewer 2: ```suggestion
-   **Up** : The default command of migrator, `up` ensures that
```

    for a given version of the migrator, every migration defined at 
    that build has been successfully applied in the connected database, 
    this is specifically important to ensure *patch version migrations are run*. 

>     ***This must be run with the target version migrator image/build**.* 

Reviewer 2: Maybe a "for example..." here showing what is meant.

Author: Gave an example, this is a tricky part of migrator 😓

    If migrator and frontend pods are deployed in version lockstep, `up` 
    ensures that ALL migrations required by the frontend will be successfully applied prior to boot.
    This is a syntactic sugar over a more internal `upto` command.

> 

Reviewer: ```suggestion
```

-   **Upgrade**: `upgrade` runs all migrations defined between two minor versions
    of Sourcegraph and requires that other services which may access the
    database are brought down. Before running, the database is checked
    for and schema drifts in order to prevent a failure while attempting
    a migration. `upgrade` relies on `stitched-migration-graph.json`.

> 

Reviewer: ```suggestion
```

-   **Drift**: This command pulls runs a diff between the current database
    schema and an expected definition packaged in migrator during the
    release process. Many migrator operations run this check before 
    proceeding to ensure the database is in the expected state.

In general, the `up` command can be thought of as a **standard** upgrade
(rolling upgrades with no downtime) while the upgrade command is what
enables **multiversion** upgrades. In part, `up` was designed to maintain
our previous upgrade policy and is thus run as an initContainer (or
initContainer-like mechanism) of the frontend, i.e. between two versions

> of Sourcegraph, the antecedent sourcegraph services will continue to

Reviewer 2: ```suggestion
of Sourcegraph, the antecedent Sourcegraph services will continue to
```

work after the consequent version's migrations have been applied.

## Current Startup Dependency

> <img src="https://storage.googleapis.com/sourcegraph-assets/Docs/migrator-startup.png" alt="Migrator Startup Dependency" className="screenshot"/>

Reviewer: ```suggestion
![Migrator Startup Dependency](https://storage.googleapis.com/sourcegraph-assets/Docs/migrator-startup.png)
```

All of our deployment types currently utilize migrator during startup.
The frontend service won't start until migrator has been run with the
default up command. The frontend service will also validate the expected
schema (and OOB migration progress), and die on startup if this
validation pass fails. This ensures that the expected migrations for the
version in question have been run.

> In docker-compose (*see diagram)*, this is accomplished via a chain of

Reviewer 2: Which diagram?

Author: This is one of the images thats loaded in, there are only two diagrams so its pretty clear when viewing a live page

> depends_on clauses in the docker-compose.yaml

Reviewer 2: ```suggestion
`depends_on` clauses in the docker-compose.yaml
```

([link](https://sourcegraph.com/github.com/sourcegraph/deploy-sourcegraph-docker/-/blob/docker-compose/docker-compose.yaml?L217-223)).

**For our k8s based deployments (including the AMIs) migrator is run as
an initContainer within the frontend utilizing the up command on the
given pods startup.**

-   [Helm Ex](https://sourcegraph.com/github.com/sourcegraph/deploy-sourcegraph-helm/-/blob/charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml?L49-76)

> 

Reviewer: ```suggestion
```

    by the frontend during an autoupgrade, displaying progress of the
    upgrade and any drift encountered.

-   All autoupgrades hit the multiversion upgrade endpoint and assume downtime for all Sourcegraph services besides the migrator and dbs.

## Migrator Release Artifacts

During the release of migrator we construct and build some artifacts
used by migrator to support its operations. Different artifacts must be
generated depending on the release type --

-   **Major**

    -   [lastMinorVersionInMajorRelease](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/oobmigration/version.go?L84-87):
        Used to evaluate what oobmigrations must run, must be updated
        every major release. This essentially tells us when a minor
        version becomes a major version. *It may be useful elsewhere at
        some point.*

-   **Minor**

    -   [maxVersionString](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@5851edf/-/blob/internal/database/migration/shared/data/cmd/generator/consts.go?L12I):

>         Defined in consts.go this string is used to tell migrator the

Reviewer 2: ```suggestion
        Defined in `consts.go` this string is used to tell migrator the
```

        latest **minor** version targetable for MVU and oobmigrations.
        [If not
        updated](https://github.com/sourcegraph/sourcegraph/issues/55048)
        multiversion upgrades cannot target the latest release. *Note
        this is used to determine how many versions should be included

>         in the stitched-migration-graph.json file.*

Reviewer 2: ```suggestion
        in the `stitched-migration-graph.json` file.*
```

    -   [Stitched-migration.json](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v5.2.0/-/blob/internal/database/migration/shared/data/stitched-migration-graph.json):
        Used by multiversion upgrades to unsquash migrations. Generated
        during release
        [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/dev/release/src/release.ts?L1101-1110).
        Learn more below.

-   **Patch**

    -   [Git_versions](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/migrator/generate.sh?L69-79):

>         Defined in generate.sh this string array contains versions of

Reviewer 2: ```suggestion
        Defined in `generate.sh` this string array contains versions of
```

        bucket (this is necessary in air gapped environments). This should
        be kept up to date with maxVersionString. [Learn
        more](https://github.com/sourcegraph/sourcegraph/issues/49813).

    -   [Squashed.sql](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v5.2.0/-/blob/migrations/frontend/squashed.sql):
        for each database we
        [generate](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/gen.sh?L18-20)
        a new squashed.sql file. It is
        [used](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v5.2.0/-/blob/internal/database/migration/drift/util_search.go?L12-31)
        to help suggest fixes for certain types of database drift. For
        example for a missing database column [this
        search](https://sourcegraph.com/search?patternType=regexp&q=repo%3A%5Egithub%5C.com%2Fsourcegraph%2Fsourcegraph%24%40v5.0.6+file%3A%5Emigrations%2Ffrontend%2Fsquashed%5C.sql%24+%28%5E%7C%5Cb%29CREATE%5CsTABLE%5Csexternal_service_sync_jobs%28%24%7C%5Cb%29+OR+%28%5E%7C%5Cb%29ALTER%5CsTABLE%5CsONLY%5Csexternal_service_sync_jobs%28%24%7C%5Cb%29&groupBy=path)
        is used to suggest a definition.

> 

Reviewer: ```suggestion
```

    -   [Schema
        descriptions](https://raw.githubusercontent.com/sourcegraph/sourcegraph/v5.2.0/internal/database/schema.json):
        schema descriptions are
        [embedded](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@v5.2.0/-/blob/cmd/migrator/generate.sh?L64-66)
        in migrator on each new build as a reference for the expected schema
        state during drift checking.

### Stitched Migration JSON and Squashing Migrations

> <img src="https://storage.googleapis.com/sourcegraph-assets/Docs/migration%20graph.png" alt="migration graph" className="screenshot"/>

Reviewer: ```suggestion
![migration graph](https://storage.googleapis.com/sourcegraph-assets/Docs/migration%20graph.png)
```

In standard/`up` upgrades `stitched-migrations.json` isn't necessary. This
is because `up` determines migrations to run by [comparing
migrations](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/internal/database/migration/definition/definition.go?L214-244)
listed as already run in the relevant db's `migration_logs` table directly
to those migration definitions
[embedded](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/migrations/embed.go)
in the migrator disk at build time for the current version, and running
any which haven't been run. We never squash away the previous minor
version of Sourcegraph, in this way we can guarantee the `migration_logs`
table migrations always has migrations in common with the migration
definitions on disk.

### Database Drifts

Database drift is any difference between the expected state of the
database at a given version and the actual state of the database. How
does it come to be? We've observed drift in customer databases for a
number of reasons, both due to admin tampering and problems in our own
release process:

-   **A backup/restore process that didn't include all objects**: This

>     notably happened to Reddit causing their database to have no

Reviewer 2: Do we want to mention a customer name here?

Author: Ahh good catch, I'll remove that

    indexes, primary keys, or unique constraints defined.

> 

Reviewer: ```suggestion
```

-   **Modifying the database explicitly**: These are untracked manual
    changes to the database schema, observed occasionally in the wild,
    and in our dotcom deployment.

> 

Reviewer: ```suggestion
```

-   **Migration failures:** that occur during multi-version upgrade or
    the up command will cause the set of *successfully applied
    migrations* to be between two versions, where drift is well-defined.

> 

Reviewer: ```suggestion
```

-   **Site Admins Error:** Errors in git-ops like deploying to
    production on the wrong version of Sourcegraph manifests have
    introduced drift. Another source is the incorrect procedure in
    downgrading.

> 

Reviewer: ```suggestion
```

 is the set of migrations defined in the `migrations` at compile time. In this way the `up` command easily facilitates dev versions of database schemas.

This "version" definition at compile time also tightly binds migrators concept of "version" to a given tag of Sourcegraph.
**The `up` command will only initialize a version of Sourcegraph, when the `migrator`
used to run `up` is the tagged version associated with the desired Sourcegraph version.**
For this reason a later version of `migrator` cannot be used to initialize an earlier
version of Sourcegraph.

#### Migration Plan

While the `up` command's concept of version is a set of embedded definitions --
the `upgrade` command does have a concept of schema migrations associated to 
version. This is the `stitched-migration-graph.json`. This file is generated on
minor releases of Sourcegraph, and defines migrations expected to have been run
at each minor version. This is necessary for two reasons --

1.  The root migration defined in the `migration` directory is a squashed 
    migration, meaning, it represents many migrations composed into a single
    sql statement.

> 

Reviewer: ```suggestion
```
